### DISCUSSION TITLE
Getting continuous disconnect/reconnect of client

### DISCUSSION THREAD

**Discussion by Drizzt321** [snapcast/snapcast#890] [Q&A]
So I'm in the process of setting up Snapcast for the first time, came across it from multiple Mopidy/Snapcast multi-room audio blog posts. 

Existing setup is multiple RPis with an amp hat and speakers which I've been using with shairport-sync, initially, for music from my phone. I also have a NAS/server, which hosts various VMs and stuff. 

Now I want to be able to have things play from a sever, both music files, and more importantly streaming from things like SoundCloud, net-radio, etc. Thus enters Mopidy, as well as Snapcast. 

So initial setup seemed fairly smooth, used 1 of the RPis and installed the snapclient onto, and installed mopidy and snapserver onto a new Debian 10.10 VM on my NAS. Got audio flowing, using Mopidy Mobile & Snapcast Android apps to control them. All good except Mopidy goes unresponsive after starting to play something, have to `kill -9` it. But that's for questions on that project.

Relevant to Snapcast, is I'm now trying to setup an AirPlay endpoint, so I or my friends can go ahead and play something to a channel for the speakers. Unfortunately that's not coming up, and there appears to be some confusion on my part between https://mjaggard.github.io/snapcast/ and https://github.com/badaix/snapcast configuration. The former (and most of the blog posts) says to put configuration in `/etc/defaults/snapserver` in the `SNAPSERVER_OPTS` variable. The main github page, here, says use `/etc/snapserver.conf`. To muddy the waters further, this post https://blog.platypush.tech/article/Build-your-open-source-multi-room-and-multi-provider-sound-server-with-Platypush-Mopidy-and-Snapcast uses `USER_OPTS` which is what I've used up until now. 

So I've tried:

```
/etc/defaults/snapserver
USER_OPTS="--user snapserver:snapserver --stream=airplay:///usr/bin/shairport-sync?name=Airplay1&devicename=Airplay1&port=5000 --stream=pipe:///tmp/snapfifo?name=mopidy&codec=pcm --codec=pcm"
```
and
```
/etc/defaults/snapserver
SNAPSERVER_OPTS="--user snapserver:snapserver --stream=airplay:///usr/bin/shairport-sync?name=Airplay1&devicename=Airplay1&port=5000 --stream=pipe:///tmp/snapfifo?name=mopidy&codec=pcm --codec=pcm"
```
and
```
/etc/snapserver.conf
[stream]
airplay:///usr/bin/shairport-sync?name=Airplay1&devicename=Airplay1&port=5000
pipe:///tmp/snapfifo?name=mopidiy&&codec=pcm
```
and the `SNAPSERVER_OPTS` version is the only one that works. So...which of `/etc/default/snapserver` with `SNAPSERVER_OPTS` or `/etc/snapserver.conf` is what should be correct and valid?

One further behavior I'm seeing, not particularly related, except insofar as I haven't been able to get the shairport-sync integration with snapserver running, I'm seeing repeated errors on the RPi client side, which causes repeating reconnection messages on the server side. 
```
Jun 25 01:06:40 kitchenmusic snapclient[535]: Exception in Controller::worker(): Can't open default PCM device: Device or resource busy
Jun 25 01:06:41 kitchenmusic snapclient[535]: Connected to 192.168.10.154
```
Which makes some sense, but it seems like the client shouldn't keep trying to reconnect to the server so much, even though it can't then open the PCM device. Seems like it's because I'm using the shairport-sync on the client device to stream some music. When I stop the audio, the messages stop. However, when I restart `snapserver`, because the client has to reconnect it starts going through that loop all over again.

It's understandable, that's why I'm trying to get the shairport-sync running on the snapserver

**Comment by kingosticks** [comment]
Do what the documentation says to do here. The people who write these blog posts often don't understand and do it wrong. Even the ones who get it right initially don't update their posts so it ends up being wrong eventually. By all means use their blogs for inspiration and ideas, but then go read the official docs. This is more work than following a step-by-step tutorial but it's got a much better chance of working. This applies to all software projects.

Some options can be put in SNAPSERVER_OPTS but most things go in /etc/snapserver.conf. Personally I'd leave SNAPSERVER_OPTS alone where possible. I would guess if it's not working it is because there's still some remenants of the wrong setup left.

**Reply by Drizzt321** [reply]
Which documentation? The one at https://github.com/badaix/snapcast, or the one at https://mjaggard.github.io/snapcast/?

So my current test is below, and I'm only getting the built-in default, without any of my configuration. And I think I just realized what it is, Debian 10.10 (buster) still has version 0.15.0. So relatively quite old. Bullseye, the next version, has 0.23.0. Might be I should just use testing (e.g. next version) on my builds, rather than putting up with the very old packages in stable. A common Debian failing *sigh*

```
/etc/snapserver.conf
[stream]
airplay:///usr/bin/shairport-sync?name=Airplay1&devicename=Airplay1&port=5000
airplay:///usr/bin/shairport-sync?name=Airplay2&devicename=Airplay2&port=5001
pipe:///tmp/snapfifo?name=mopidiy&&codec=pcm

/etc/default/snapserver
SNAPSERVER_OPTS=""
```

**Reply by kingosticks** [reply]
The documentation here in this repository at https://github.com/badaix/snapcast. The other one seems to be some random guy's random fork. 

I kinda assumed you had the latest version. If you choose Debian you choose stability. If you want the latest version just take the deb from https://github.com/badaix/snapcast/releases/latest

**Reply by Drizzt321** [reply]
Gotcha, ok. Even though that other is the top DuckDuckGo (and thus Bing) result *sigh*  Maybe reach out to him and ask him to take it down, since it's confusing people?

And yeah, I'm just going to upgrade to the `testing` target, and take my chances. Should be pretty stable in general, and keep me reasonably up to date. Thanks!

Although that disconnect/reconnect filling up the logs seems like it'd still exist, yet I can't think of a better way to be sure things are properly logged.

**Reply by Drizzt321** [reply]
Scratch all that. I had some .conf file syntax incorrect.
